Fix annotations guarding ServiceWorkersEnabled endpoints

darryl-apple opened the pull request:
#### 095c9dd02d7f84891cf4a3d00063343c4de0dc97
<pre>
Fix annotations guarding ServiceWorkersEnabled endpoints
<a href="https://bugs.webkit.org/show_bug.cgi?id=296802">https://bugs.webkit.org/show_bug.cgi?id=296802</a>
<a href="https://rdar.apple.com/157291144">rdar://157291144</a>

Reviewed by Sihui Liu.

Annotate endpoints with ServiceWorkersEnabled
Prevent incorrect annotations from being used in .messages.in files.

* LayoutTests/http/wpt/push-api/pushEvent.any.serviceworker.html:
* LayoutTests/http/wpt/push-api/pushManager.any.html:
* LayoutTests/http/wpt/push-api/pushManager.any.serviceworker.html:
* LayoutTests/http/wpt/push-api/pushSubscription.https.any.html:
* LayoutTests/http/wpt/push-api/pushSubscription.https.any.serviceworker.html:
* LayoutTests/http/wpt/push-api/pushSubscriptionChangeEvent.any.serviceworker.html:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.cpp:
(WebKit::ServiceWorkerDownloadTask::ServiceWorkerDownloadTask):
(WebKit::ServiceWorkerDownloadTask::sharedPreferencesForWebProcess const):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.h:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.messages.in:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.cpp:
(WebKit::ServiceWorkerFetchTask::sharedPreferencesForWebProcess const):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.h:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.messages.in:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.messages.in:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
(WebKit::WebSWServerToContextConnection::sharedPreferencesForWebProcess const):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.messages.in:
* Source/WebKit/Scripts/webkit/parser.py:
(parse):
(parse_coalescing_keys):
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:

Canonical link: <a href="https://commits.webkit.org/298775@main">https://commits.webkit.org/298775@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/721b3c03a6b632f8e0a98c29c7b430ce0b7e7530

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116589 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36253 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26815 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122661 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67160 "Built successfully") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36944 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44848 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88552 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42982 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2b1b5a30-4b23-4f48-819c-6f6063eb77a3) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119538 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29471 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104601 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69012 "Passed tests") | | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/115963 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28537 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22706 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66328 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98852 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22866 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125799 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43488 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32674 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97223 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43852 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100803 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97015 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24706 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42329 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20266 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39450 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43371 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48974 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42840 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46180 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44546 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.messages.in:
```diff
@@ -27,33 +27,33 @@
     # When possible, these messages can be implemented directly by WebCore::SWClientConnection
-    [EnabledBySetting=ServiceWorkersEnabled] ScheduleJobInServer(struct WebCore::ServiceWorkerJobData jobData)
-    [EnabledBySetting=ServiceWorkersEnabled] ScheduleUnregisterJobInServer(WebCore::ServiceWorkerJobIdentifier jobIdentifier, WebCore::ServiceWorkerRegistrationIdentifier identifier, WebCore::ServiceWorkerOrClientIdentifier documentIdentifier) -> (Expected<bool, WebCore::ExceptionData> result)
+    [EnabledBy=ServiceWorkersEnabled] ScheduleJobInServer(struct WebCore::ServiceWorkerJobData jobData)
```
I am surprised "EnabledBySetting" is a supported attribute for messages.in files; is it something wrong in the parser?
And you might need to drop `ExceptionForEnabledBy` in the file if all messages are annotated

darryl-apple replied:
Yep me too, I've changed the parser to only allow certain attributes

szewai commented on Source/WebKit/Scripts/webkit/parser.py:
```diff
@@ -61,6 +61,8 @@ def parse(file):
     match = re.search(r'\s*\[\s*(?P<extended_attributes>.*?)\s*\]\s*.*messages -> ', "".join(file_contents).replace("\n", " "), re.MULTILINE)
     if match:
         extended_attributes = re.split(r'\s*,\s*', match.group('extended_attributes'))
+        validate_receiver_annotations(extended_attributes)
```
Why not just throw error if the text does not match any pattern, e.g.:
```
if match:
    if match.group('name') == 'EnabledBy':
        (receiver_enabled_by, receiver_enabled_by_conjunction) = parse_enabled_by_string(match.group('value'))
        continue
    if match.group('name') == 'DispatchedFrom':
        receiver_dispatched_from = parse_process_name_string(match.group('value'))
        continue
    if match.group('name') == 'DispatchedTo':
        receiver_dispatched_to = parse_process_name_string(match.group('value'))
        continue
    raise Exception("ERROR: Unknown extended attribute")
elif attribute == 'SharedPreferencesNeedsConnection':
    ...
```

darryl-apple replied:
This is a great suggestion, changed.

szewai commented on Source/WebKit/Scripts/webkit/parser.py:
```diff
@@ -128,6 +129,8 @@ def parse(file):
             enabled_by_conjunction = None
             coalescing_key_indices = None
             if options_string:
+                validate_message_annotations(options_string, name)
```
Can we throw error after the line below?
```
if match:
    coalescing_key_indices = parse_coalescing_keys(match.group(1), [parameter.name for parameter in parameters])
else:
    raise Exception("ERROR: Unknown extended attribute '%s'" % attribute)
```

darryl-apple replied:
no I don't think so. match is reset with each re.search, so this would only check the result of 
```
 match = re.search(r"(?:(?:, |^)+(?:DeferSendingIfSuspendedWithCoalescingKeys=\((.*?)\)))(?:, |$)?", options_string)
```
The problem is that if there's multiple annotations then several of the match flags get set, however this doesn't necessarily match the number of annotations.

darryl-apple replied:
Maybe I can change change the code to use a similar approach as the interface level checks rather than using regex

szewai reviewed (approved):
r=me if bots are happy

darryl-apple committed:
Fix annotations guarding ServiceWorkersEnabled endpoints
https://bugs.webkit.org/show_bug.cgi?id=296802
rdar://157291144

Reviewed by Sihui Liu.

Annotate endpoints with ServiceWorkersEnabled
Prevent incorrect annotations from being used in .messages.in files.

* LayoutTests/http/wpt/push-api/pushEvent.any.serviceworker.html:
* LayoutTests/http/wpt/push-api/pushManager.any.html:
* LayoutTests/http/wpt/push-api/pushManager.any.serviceworker.html:
* LayoutTests/http/wpt/push-api/pushSubscription.https.any.html:
* LayoutTests/http/wpt/push-api/pushSubscription.https.any.serviceworker.html:
* LayoutTests/http/wpt/push-api/pushSubscriptionChangeEvent.any.serviceworker.html:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.cpp:
(WebKit::ServiceWorkerDownloadTask::ServiceWorkerDownloadTask):
(WebKit::ServiceWorkerDownloadTask::sharedPreferencesForWebProcess const):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.h:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.messages.in:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.cpp:
(WebKit::ServiceWorkerFetchTask::sharedPreferencesForWebProcess const):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.h:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.messages.in:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.messages.in:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
(WebKit::WebSWServerToContextConnection::sharedPreferencesForWebProcess const):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.messages.in:
* Source/WebKit/Scripts/webkit/parser.py:
(parse):
(parse_coalescing_keys):
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:

Canonical link: https://commits.webkit.org/298775@main
